source/subscription-manager/subscription_manager.h

```diff
@@ -0,0 +1,124 @@
+/*
+ * FreeRTOS V202011.00
+ * Copyright (C) 2020 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * https://www.FreeRTOS.org
+ * https://aws.amazon.com/freertos
+ *
+ */
+
+/**
+ * @file subscription_manager.h
+ * @brief Functions for managing MQTT subscriptions.
+ */
+#ifndef SUBSCRIPTION_MANAGER_H
+#define SUBSCRIPTION_MANAGER_H
+
+/* Demo config include. */
+#include "demo_config.h"
+
+/* Kernel includes. */
+#include "FreeRTOS.h"
+
+/* core MQTT include. */
+#include "core_mqtt.h"
+
+
+/**
+ * @brief Maximum number of subscriptions maintained by the subscription manager
+ * simultaneously in a list.
+ */
+#ifndef SUBSCRIPTION_MANAGER_MAX_SUBSCRIPTIONS
+    #define SUBSCRIPTION_MANAGER_MAX_SUBSCRIPTIONS    10U
+#endif
+
+/**
+ * @brief Callback function called when receiving a publish.
+ *
+ * @param[in] pvIncomingPublishCallbackContext The incoming publish callback context.
+ * @param[in] pxPublishInfo Deserialized publish information.
+ */
+typedef void (* IncomingPubCallback_t )( void * pvIncomingPublishCallbackContext,
+                                         MQTTPublishInfo_t * pxPublishInfo );
+
+/**
+ * @brief An element in the list of subscriptions maintained by the agent.
+ *
+ * @note The agent allows multiple tasks to subscribe to the same topic.
+ * In this case, another element is added to the subscription list, differing
+ * in the intended publish callback.
+ */
+typedef struct subscriptionElement
+{
+    IncomingPubCallback_t pxIncomingPublishCallback;
+    void * pvIncomingPublishCallbackContext;
+    uint16_t uFilterStringLength;
```

Comment:
If we're following the rest of the style, shouldn't this be labeled as an unsigned short?
```suggestion
    uint16_t usFilterStringLength;
```